### PR TITLE
allow xform backup pillow

### PR DESCRIPTION
**Pull request by calellowitz**
## Technical Summary

This allows the index for the xform pillow to be set dynamically, similar to how it has already been implemented for the case search pillow. It will have no effect until a pillow using that constructor is added in localsettings.

### Safety story

This constructor is not currently in use, but should be safe when it is in use because it will only affect the backfill pillow, and uses a well understood process.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/pillows/xform.py [hunk 1]
```diff
@@ -143,9 +143,22 @@ def get_xform_to_elasticsearch_pillow(pillow_id='XFormToElasticsearchPillow', nu
...
       - :py:class:`pillowtop.processors.elastic.ElasticProcessor`
     """
+    index_info = XFORM_INDEX_INFO
     # todo; To remove after full rollout of https://github.com/dimagi/commcare-hq/pull/21329/
-    assert pillow_id == 'XFormToElasticsearchPillow', 'Pillow ID is not allowed to change'
-    checkpoint = get_checkpoint_for_elasticsearch_pillow(pillow_id, XFORM_INDEX_INFO, FORM_TOPICS)
+    if 'index_name' in kwargs:
+        raw_info = XFORM_INDEX_INFO.to_json()
+        raw_info.pop("meta")
+        index_info = ElasticsearchIndexInfo.wrap(raw_info)
```
Comment (dannyroberts) [inline corehq/pillows/xform.py:151]: Is the stickler here right that this needs to be imported?

Comment (calellowitz) [reply corehq/pillows/xform.py:151]: Apparently

**Review by snopoke — APPROVED** [review]

**Review by joxl — COMMENTED** [review]
In my opinion it would be helpful to consider adding tests around how the new pillow works. For example, what is the expected behavior with the `pillow_retry` process? Are retries applied to both indexes, or just the one the error occurred on?

File: corehq/pillows/xform.py [hunk 2]
```diff
@@ -143,9 +144,22 @@ def get_xform_to_elasticsearch_pillow(pillow_id='XFormToElasticsearchPillow', nu
...
+        index_info.index = kwargs['index_name']
+        index_info.alias = kwargs['index_alias']
+        from corehq.apps.es.registry import register
```
[the remark this answers is not in the document]
Comment (calellowitz) [reply corehq/pillows/xform.py]: Thanks for pointing that out. My read of the new state of the code and local testing is that this is now unnecessary and the `add_dynamic_adapter` is sufficient, so I have removed this reference entirely, but let me know if that does not match your understanding.

**Comment by calellowitz** [comment]
> In my opinion it would be helpful to consider adding tests around how the new pillow works. For example, what is the expected behavior with the `pillow_retry` process? Are retries applied to both indexes, or just the one the error occurred on?

Can you explain a bit more what you would like me to add testing? This uses the normal pillow retry logic, which tracks the pillow on which the error occurs. This pillow only knows about one index, so any errors that occur in this pillow will only be retried on the configured index, the index it writes to. It will not touch the current form index at all. There was a bug in this PR earlier, where I missed a reference to the other index that I needed to update. Apologies if that is what was causing the confusion.

**Comment by joxl** [comment]
> any errors that occur in this pillow will only be retried on the configured index, the index it writes to. It will not touch the current form index at all.

I see. Eventually that index will be swapped and become the (only) "current form index", but during the backfill operation, retry records can get created for two different indexes. Test cases for this data processing flow is what I was referring to.  So something like `test_backfill_index_retry_records_get_retried_on_main_index_after_swap()` and `test_main_index_retry_records_<some_action>_after_swap()`, etc.

**Comment by calellowitz** [comment]
I took another look at the pillow retry code, and I believe my previous explanation was not correct. While the errors are tagged by pillow, retry code for kafka pillows only republishes the change, where it will then be picked up by all pillows consuming from that kafka topic. So if any errors occur in the backfill pillow, they will be republished in kafka and be picked up by both pillows. That means that as long as the backfill pillow is running it will retry any of its own errors, and when we update settings, and the main pillow begins writing to the new index, it will continue to retry up errors from the backfill pillow, since they will consume from the same topic.

The previous explanation I gave, using the pillow itself to retry the change, is true for couch pillows, but not for kafka ones.

**Comment by joxl** [comment]
Thanks for the explanation. In my opinion, tests would be really helpful for reviewing these expectations.

**Comment by calellowitz** [comment]
> Thanks for the explanation. In my opinion, tests would be really helpful for reviewing these expectations.

Can you explain a bit more about the tests you are looking for? This PR does not change any logic around pillow retry or pillow consumers, and the code to process the retries (by republishing to kafka) and consume from kafka topics (as both this pillow and the existing form pillow do) are well tested already. In fact, this specific pillow is already tested in pillow retry code as part of the [all pillows test](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/pillow_retry/tests/test_all_pillows.py#L28) since it is included in our [settings file](https://github.com/dimagi/commcare-hq/blob/master/settings.py#L1705)

**Review by snopoke — APPROVED** [review]
On the testing topic, if this workflow were the long term plan for reindexing then it would be good to have a test that verified (and documented) that an error generated by one pillow is picked up by another consuming from the same Kafka topic. (logically we can see from the design that it will be but it is useful to have that documented with a test). However since this workflow is soon to be deprecated, and since we have used it recently without issue I don't think such tests have much value. If we still want to add them we can but I wouldn't want them to hold up this PR and the current reindex process that is underway.

**Comment by calellowitz** [comment]
I offlined with @joxl and his feedback is non blocking, so in an effort to finish this as quickly as possible, I am going to merge now.

**Pull request merged**